File: Data Structures & Algorithms/longest-increasing-subsequence/submission-0.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        if not nums:
            return 0
        n = len(nums)
        dp = [1] * n  # dp[i] = length of LIS that ends at i
        for i in range(n): # The shifting window on the right
            for j in range(i): # Consider all combinations 
                if nums[j] < nums[i]:
                    dp[i] = max(dp[i], dp[j] + 1)
        return max(dp)

File: Data Structures & Algorithms/jump-game/submission-0.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        if (nums.empty()) return false;
        int closestJumpIndex = nums.size() - 1;
        for (int i = nums.size() - 2; i >= 0; --i) {
            if (nums[i] + i >= closestJumpIndex) {
                closestJumpIndex = i;
            }
        }
        return closestJumpIndex == 0;
    }
};
